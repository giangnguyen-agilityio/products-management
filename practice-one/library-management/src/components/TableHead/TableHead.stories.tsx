import {type StoryFn, type Meta} from '@storybook/react'
import TableHead from './index'
import React from 'react'

export default {
  title: 'Example/TableHead',
  component: TableHead,
} as Meta<typeof TableHead>

const Template: StoryFn<typeof TableHead> = args => <TableHead {...args} />
export const Default = Template.bind({})
Default.args = {
  columns: [
    {field: 'id', headerName: 'ID'},
    {field: 'bookId', headerName: 'Book ID'},
    {field: 'memberId', headerName: 'Member ID'},
    {field: 'fromDate', headerName: 'From Date'},
    {field: 'toDate', headerName: 'To Date'},
    {field: 'status', headerName: 'Status'},
  ],
}
