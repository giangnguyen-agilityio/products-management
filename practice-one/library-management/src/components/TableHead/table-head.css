.table-header {
  margin: 15px 0;
  background: var(--bg-introduction-color);
  color: var(--secondary-text-color);
}

.table-header-cell {
  text-align: center;
  padding: 15px;
  width: 25%;
}

.table-header-cell:first-child,
.table-header-cell:nth-child(2),
.table-header-cell:nth-child(3) {
  width: 15%;
}
