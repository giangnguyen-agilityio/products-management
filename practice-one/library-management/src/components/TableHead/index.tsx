import React from 'react'
import {TableColumn} from '@types'
import {ROLE} from '@constants'
import {getItemInLocalStorage} from '@helpers'
import './table-head.css'

interface TableHeadProps {
  columns: TableColumn[]
}

const TableHead: React.FC<TableHeadProps> = props => {
  const {columns} = props

  const isAdmin: boolean = getItemInLocalStorage('memberRole') === ROLE.ADMIN

  return (
    <thead className="table-header">
      <tr className="table-header-row">
        {/* Render table header cells for each column */}
        {columns.map(column => (
          <th key={column.field} className="table-header-cell">
            {/* Display the header name of the column */}
            {column.headerName}
          </th>
        ))}
        {/* Render the "Action" header cell */}
        {isAdmin ? <th className="table-header-cell">Action</th> : null}
      </tr>
    </thead>
  )
}

export default TableHead
