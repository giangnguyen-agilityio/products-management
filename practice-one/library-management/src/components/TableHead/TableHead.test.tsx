import React from 'react'
import {render} from '@testing-library/react'
import * as helpers from '@helpers'
import TableHead from '.'

jest.mock('@helpers', () => ({
  getItemInLocalStorage: jest.fn(),
}))

describe('TableHead component', () => {
  const columns = [
    {field: 'column1', headerName: 'Column 1'},
    {field: 'column2', headerName: 'Column 2'},
    {field: 'column3', headerName: 'Column 3'},
    {field: 'column4', headerName: 'Column 4'},
  ]

  const tableElement = document.createElement('table')

  it('should renders table header cells correctly', () => {
    const {container, getByText} = render(<TableHead columns={columns} />, {
      container: document.body.appendChild(tableElement),
    })

    columns.forEach(column => {
      const columnHeader = getByText(column.headerName)
      expect(columnHeader).toBeInTheDocument()
    })

    expect(container).toMatchSnapshot()
  })

  it('should renders the "Action" header cell when user is an Admin', () => {
    const getItemInLocalStorageMock = jest.spyOn(
      helpers,
      'getItemInLocalStorage'
    )
    getItemInLocalStorageMock.mockReturnValue('admin')

    // Render the component with the desired props
    const {container, getByText} = render(<TableHead columns={columns} />, {
      container: document.body.appendChild(tableElement),
    })

    // Assert that the "Action" header cell is rendered
    expect(getByText('Action')).toBeInTheDocument()

    expect(container).toMatchSnapshot()
  })

  it('should does not render the "Action" header cell when user is not an Admin', () => {
    const getItemInLocalStorageMock = jest.spyOn(
      helpers,
      'getItemInLocalStorage'
    )
    getItemInLocalStorageMock.mockReturnValue(null)

    const {container, queryByText} = render(<TableHead columns={columns} />, {
      container: document.body.appendChild(tableElement),
    })
    expect(queryByText('Action')).toBeNull()

    expect(container).toMatchSnapshot()
  })
})
