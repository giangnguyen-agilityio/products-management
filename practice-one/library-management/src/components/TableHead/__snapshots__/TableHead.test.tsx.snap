// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableHead component should does not render the "Action" header cell when user is not an Admin 1`] = `
<table>
  <thead
    class="table-header"
  >
    <tr
      class="table-header-row"
    >
      <th
        class="table-header-cell"
      >
        Column 1
      </th>
      <th
        class="table-header-cell"
      >
        Column 2
      </th>
      <th
        class="table-header-cell"
      >
        Column 3
      </th>
      <th
        class="table-header-cell"
      >
        Column 4
      </th>
    </tr>
  </thead>
</table>
`;

exports[`TableHead component should renders table header cells correctly 1`] = `
<table>
  <thead
    class="table-header"
  >
    <tr
      class="table-header-row"
    >
      <th
        class="table-header-cell"
      >
        Column 1
      </th>
      <th
        class="table-header-cell"
      >
        Column 2
      </th>
      <th
        class="table-header-cell"
      >
        Column 3
      </th>
      <th
        class="table-header-cell"
      >
        Column 4
      </th>
    </tr>
  </thead>
</table>
`;

exports[`TableHead component should renders the "Action" header cell when user is an Admin 1`] = `
<table>
  <thead
    class="table-header"
  >
    <tr
      class="table-header-row"
    >
      <th
        class="table-header-cell"
      >
        Column 1
      </th>
      <th
        class="table-header-cell"
      >
        Column 2
      </th>
      <th
        class="table-header-cell"
      >
        Column 3
      </th>
      <th
        class="table-header-cell"
      >
        Column 4
      </th>
      <th
        class="table-header-cell"
      >
        Action
      </th>
    </tr>
  </thead>
</table>
`;
